### ISSUE TITLE
VS Code cannot bind breakpoints in a Turborepo Next.js project

### ISSUE THREAD

**Issue by michaelschufi** [vercel/next.js#73258]
### Link to the code that reproduces this issue

https://github.com/michaelschufi/turbo-debug-nextjs-repro

### To Reproduce

1. Using VS Code, open either 
   the file `turbo-debug-nextjs-repro.code-workspace`:
   ```
   code turbo-debug-nextjs-repro.code-workspace
   ```
   
   or the folder `apps/docs/`:
   ```
   code apps/docs
   ```
2. Open the file `app/page.tsx` inside the folder `apps/docs/`.
3. Set a breakpoint on the same line as the `console.log`.
4. Start the debugger and open the page in a browser.

### Current vs. Expected behavior

Observe
- the debugger NOT stopping at the breakpoint
   - the debugger stopping at the debugger statement
   - the debugger opening a second VS Code file tab
   - the paths of the two tabs not matching when hovering over the tab labels (see screenshot below)  

-> As can be seen: The debugger kind of adds the relative path to the workspace folder to the path of the file it is debugging - resulting in doubling the path segments.

```diff
- ~/repos/turbo-debug-nextjs-repro/apps/docs/app/page.tsx
+ ~/repos/turbo-debug-nextjs-repro/apps/docs/apps/docs/app/page.tsx
```

### Provide environment information

```bash
Operating System:
  Platform: linux
  Arch: x64
  Version: #1 SMP Fri Mar 29 23:14:13 UTC 2024
  Available memory (MB): 15853
  Available CPU cores: 16
Binaries:
  Node: 22.11.0
  npm: 10.9.0
  Yarn: N/A
  pnpm: 9.0.0
Relevant Packages:
  next: 15.0.3 // Latest available version is detected (15.0.3).
  eslint-config-next: N/A
  react: 19.0.0-rc-66855b96-20241106
  react-dom: 19.0.0-rc-66855b96-20241106
  typescript: 5.5.4
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Developer Experience, Turbopack

### Which stage(s) are affected? (Select all that apply)

next dev (local)

### Additional context

I think, this might also be related to the Next.js dev error window's path not matching the editor's CWD - instead it also outputs the path from the monorepo-root instead of the CWD where `next dev` was called.

Note: I don't know if this is a Next.js, Turborepo, Turbopack, or a VS Code issue - probably a bit of all of them 😅

**Comment by emilioestebanez** [comment]
I think this is the same thing as https://github.com/vercel/next.js/issues/73214

**Comment by samcx** [comment]
@michaelschufi Closing in favor of the other one mentioned above!

**Comment by michaelschufi** [comment]
Hmm, I'm not sure. For me at least, the breakpoint in #73214's repo binds fine.
Also, this issue is about the breakpoints not binding in a Turborepo monorepo because Next.js seems to add the relative path to the workspace folder to the path of the file it is debugging - resulting in doubling the path segments.

**Comment by Falven** [comment]
> Hmm, I'm not sure. For me at least, the breakpoint in #73214's repo binds fine. Also, this issue is about the breakpoints not binding in a Turborepo monorepo because Next.js seems to add the relative path to the workspace folder to the path of the file it is debugging - resulting in doubling the path segments.
> 
> Edit: It's definitely a different issue @samcx (see [#73214 (comment)](https://github.com/vercel/next.js/issues/73214#issuecomment-2505965622)). The other issue is about a breakpoint not binding on Windows, whereas for this one, all platforms are affected by the current behaviour.

This is definitely an issue. I posted a temporary workaround here https://github.com/vercel/turborepo/issues/9579 the problem with this workaround is it breaks every other form of debugging e.g. if you want to debug your package from your tests… so I’m looking for some better guidance, maybe some way to modify the source roots only from the NextJS app with web pack.

**Comment by Falven** [comment]
> > Hmm, I'm not sure. For me at least, the breakpoint in #73214's repo binds fine. Also, this issue is about the breakpoints not binding in a Turborepo monorepo because Next.js seems to add the relative path to the workspace folder to the path of the file it is debugging - resulting in doubling the path segments.
> > Edit: It's definitely a different issue @samcx (see [#73214 (comment)](https://github.com/vercel/next.js/issues/73214#issuecomment-2505965622)). The other issue is about a breakpoint not binding on Windows, whereas for this one, all platforms are affected by the current behaviour.
> 
> This is definitely an issue. I posted a temporary workaround here [vercel/turborepo#9579](https://github.com/vercel/turborepo/issues/9579) the problem with this workaround is it breaks every other form of debugging e.g. if you want to debug your package from your tests… so I’m looking for some better guidance, maybe some way to modify the source roots only from the NextJS app with web pack.

Oh and I think this has been an issue for a long time. I recall another fix I did a while back was to do cd ‘../.. && next dev’ (or w/e) the next command was in the debug command. You could try that.

**Comment by Falven** [comment]
FYI I scoured the Next.JS source code and found these comments in `packages/next/src/build/webpack-config.ts` which are directly relevant to the issue at hand:

```ts
// if `sources[number]` is not an absolute path, it's is resolved
// relative to the location of the source map file (https://tc39.es/source-map/#resolving-sources).
// However, Webpack's `resource-path` is relative to the app dir.
// TODO: Either `sourceRoot` should be populated with the root and then we can use `[resource-path]`
// or we need a way to resolve return `path.relative(sourceMapLocation, info.resourcePath)`
```

**What this means:**

- The official ECMAScript spec says that if paths in sources are not absolute after sourceRoot is applied, they must be resolved relative to the .map file's location.
- Next.js currently doesn't set a sourceRoot for modules in node_modules. Instead, it uses [resource-path], which is relative to the application directory, not the package in node_modules.
- Without a proper sourceRoot, relative paths do not resolve as intended, leading to the debugger failing to locate the correct source files.

**In short:** The bug is that Next.js never sets sourceRoot for these module source maps, causing them to be resolved from the wrong directory. To fix this, Next.js needs to:

- Set sourceRoot for each module's source maps to the actual root of the module.
- Alternatively, ensure that the devtoolModuleFilenameTemplate and resource-path usage respects the module's location and aligns with how the ECMAScript standard dictates source resolution.

**Comment by michaelschufi** [comment]
@Falven I think the problem happens even without any other internal workspace dependencies. 
Could the issue be due to Next.js assuming the `node_modules` root folder is always in Next.js's project folder and not above, like in a monorepo with symlinks?

**Comment by AndreMaz** [comment]
Had the same issue with my project that uses [t3-oss](https://github.com/t3-oss/create-t3-app) template.

After bumping next to v15.2.0-canary.8 and turbo to v2.3.3 my debugger started to work again

**Comment by michaelschufi** [comment]
I'm closing this, as the original issue (the path one) seems to be resolved. Thanks!

However, I still experience problems binding breakpoints in some cases. I will open a new issue if I can find a reproducible case.

**Comment by Godrules500** [comment]
I'm having the same issue. VScode won't hit breakpoints when using turbo.
